src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -876,284 +876,170 @@ Int64 ZooKeeper::getClientID()
     return impl->getSessionID();
 }
 
-
-std::future<Coordination::CreateResponse> ZooKeeper::asyncCreate(const std::string & path, const std::string & data, int32_t mode)
+template <class T>
+auto makeContract()
 {
-    /// https://stackoverflow.com/questions/25421346/how-to-create-an-stdfunction-from-a-move-capturing-lambda-expression
-    auto promise = std::make_shared<std::promise<Coordination::CreateResponse>>();
+    auto promise = std::make_shared<std::promise<T>>();
     auto future = promise->get_future();
-
-    auto callback = [promise, path](const Coordination::CreateResponse & response) mutable
-    {
-        if (response.error != Coordination::Error::ZOK)
-            promise->set_exception(std::make_exception_ptr(KeeperException(path, response.error)));
-        else
-            promise->set_value(response);
-    };
-
-    impl->create(path, data, mode & 1, mode & 2, {}, std::move(callback));
-    return future;
+    return std::make_pair(std::move(promise), std::move(future));
 }
 
-std::future<Coordination::CreateResponse> ZooKeeper::asyncTryCreateNoThrow(const std::string & path, const std::string & data, int32_t mode)
+template <class T>
+auto makeNothrowCallback(std::shared_ptr<std::promise<T>> promise)
 {
-    auto promise = std::make_shared<std::promise<Coordination::CreateResponse>>();
-    auto future = promise->get_future();
+    return [promise = std::move(promise)](const T& response) mutable { promise->set_value(response); };
+}
 
-    auto callback = [promise](const Coordination::CreateResponse & response) mutable
+template <Coordination::Error ...errs, class T>
```

Comment:
I like some tiny copy-paste more than variadic templates for error codes.

Reply:
Well, this is some sort of black box that does the job. On the contrary, I believe that this is a better solution than manually repeating similar lambdas.